File: ClearFiveStages-HTML5/CFSGamePlayScene.js
var ROW_CLEARED_LABEL = 6000;
var WAITING_LABEL = 7000;

// init
var CFSGamePlayScene = function () {
};

CFSGamePlayScene.prototype.onDidLoadFromCCB = function () {
    this.tabelBgSprite.setTexture(cc.TextureCache.getInstance().addImage("tableBg{0}.jpg".format(GameDataManager.tableBgIndex)));
	this.initData();
	this.disableFunctionMenu();
	this.disablePlacePaiGowMenu();
	// update elements for multi-language support
	this.updateCCBElements();
    // schedule
    this.rootNode.onTimerUpdate = function(dt) {
        this.controller.updateTimeString();
    };
    this.rootNode.onTouchesBegan = function(touches, event) {
        this.controller.onTouchesBegan(touches, event);
        return true;
    };
    if (!this.isPaused) {
        CFSModePaiGowManager.getInstance().initAllPowGowObjects();
        this.runReadySetGoAnimation();
    }
};

CFSGamePlayScene.prototype.initData = function () {
	this.timeCount = 0;
	this.placePaiGowRowNum = 999;
	this.cfsScoreData = new CFSScoreData();
	this.isPaused = false;
	this.isGameOver = false
	this.cfsScoreData.playTimeInSec = 0;
	this.cfsScoreData.hitPercentage = 0.0;
	GameDataManager.resetHintCount();
	CFSModePaiGowManager.getInstance().tableLayer = this.rootNode;
	CFSModePaiGowManager.getInstance().delegate = this;
	this.rootNode.setTouchEnabled(false);
};

CFSGamePlayScene.prototype.onEnter = function () {
	if (!this.isPaused && !this.isGameOver) {
		CFSModePaiGowManager.getInstance().initAllPowGowObjects();
		this.runReadySetGoAnimation();
	}
};

CFSGamePlayScene.prototype.updateCCBElements = function () {
    this.dealCardButtonLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_DEALCARD));
    this.nextRowButtonLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_NEXTROW));
    this.getCardButtonLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_GETCARD));
    this.hintButtonLabel.setString("{0}({1})".format(MultiLanguageUtil.getLocalizatedStringForKey(STRING_SHOWHINT), GameDataManager.hintCount));
    this.timeLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_TIMER));
};

// touch
CFSGamePlayScene.prototype.onTouchesBegan = function(touches, event) {
	CFSModePaiGowManager.getInstance().handleTouch(touches[0]);
};

// function menu
CFSGamePlayScene.prototype.enableFunctionMenu = function () {
    this.dealMenuItem.setEnabled(true);
    this.getMenuItem.setEnabled(true);
    this.nextMenuItem.setEnabled(true);
    this.pauseMenuItem.setEnabled(true);
    if (GameDataManager.hintCount > 0) this.hintMenuItem.setEnabled(true);
};

CFSGamePlayScene.prototype.disableFunctionMenu = function () {
    this.dealMenuItem.setEnabled(false);
    this.getMenuItem.setEnabled(false);
    this.nextMenuItem.setEnabled(false);
    this.hintMenuItem.setEnabled(false);
    this.pauseMenuItem.setEnabled(false);
};

CFSGamePlayScene.prototype.enablePlacePaiGowMenu = function () {
    for (var rowNum = 1; rowNum <= 5; rowNum ++) {
        if (CFSModePaiGowManager.getInstance().isRowCleared(rowNum)) continue;
        var arrowSprite = new cc.Sprite();
        arrowSprite.initWithSpriteFrameName("arrowL.png");
        arrowSprite.setScale(0.7);
        var arrowMenuItem = cc.MenuItemSprite.create(arrowSprite, null, null, this.onPlacePaiGowPressed, this);
        arrowMenuItem.setTag(rowNum);
        var menu = cc.Menu.create(arrowMenuItem, null);
        var paiGowObject = CFSModePaiGowManager.getInstance().anyPaiGowForRow(rowNum);
        menu.setPosition(cc.p(cc.Director.getInstance().getWinSize().width, paiGowObject.pg_position.y));
        this.rootNode.addChild(menu, 1, 80000 + rowNum);
        menu.runAction(cc.RepeatForever.create(cc.JumpBy.create(1.0, cc.p(0, 0), 10, 1)));
    }
};

CFSGamePlayScene.prototype.disablePlacePaiGowMenu = function () {
    for (var rowNum = 1; rowNum <= 5; rowNum ++) this.rootNode.removeChildByTag(80000 + rowNum);
};

// animation
CFSGamePlayScene.prototype.runReadySetGoAnimation = function () {
    var winSize = cc.Director.getInstance().getWinSize();
    var readySetGoNode =  cc.BuilderReader.load("ReadySetGo.ccbi");
    readySetGoNode.setPosition(cc.p(winSize.width / 2, winSize.height / 2));
    this.rootNode.addChild(readySetGoNode, 999);
    readySetGoNode.animationManager.runAnimationsForSequenceNamed("ReadySetGo");
    readySetGoNode.animationManager.setCompletedAnimationCallback(this, this.removeSpriteAndBegin);
    CFSModePaiGowManager.getInstance().putAllPaiGowsOnTable();
    CFSModePaiGowManager.getInstance().turnAroundAllPaiGowRowsWithInterval(0.5);
};

// update UI
CFSGamePlayScene.prototype.updateTimeString = function () {
    this.timeCount ++;
    this.timeLabel.setString(MultiLanguageUtil.getTimeStringFromSecond(this.timeCount));
};

CFSGamePlayScene.prototype.setInfoString = function (info) {
	this.gameTipLabel.stopAllActions();
	this.gameTipLabel.setString(info);
	this.gameTipLabel.setOpacity(255);
	this.gameTipLabel.runAction(cc.Sequence.create(cc.Blink.create(1.0, 5), cc.FadeOut.create(0.5)));
};

CFSGamePlayScene.prototype.removeBubForRow = function (rowNum) {
    this.rootNode.removeChildByTag(rowNum, true);
};

CFSGamePlayScene.prototype.createBubForRow = function (rowNum, toHide, bubPosition) {
    this.rootNode.removeChildByTag(rowNum, true);
    var bub = new cc.Sprite();
    bub.initWithSpriteFrameName("bub.png");
    var bubInfoLabel = cc.LabelTTF.create("×{0}".format(toHide), "Verdana-BoldItalic", 15);
    bubInfoLabel.setColor(cc.black);
    bubInfoLabel.setPosition(cc.p(25, 25));
    bub.addChild(bubInfoLabel, 1);
    var menuItem = cc.MenuItemSprite.create(bub, null, null, this.showHiddenPaiGow, this);
    menuItem.setTag(rowNum);
    menuItem.setScale(0.7);
    var menu = cc.Menu.create(menuItem, null);
    menu.setPosition(bubPosition);
    this.rootNode.addChild(menu, 3, rowNum);
    menu.runAction(cc.RepeatForever.create(cc.JumpBy.create(1.5, cc.p(0, 0), 6, 1)));
};

CFSGamePlayScene.prototype.showHiddenPaiGow = function (sender) {
    if (this.rootNode.getChildByTag(400 + sender.getTag()) == null) {
        var layer = ShowPaiGowLayer.create(sender.getTag());
        this.rootNode.addChild(layer, 999, 400 + sender.getTag());
    }
};

CFSGamePlayScene.prototype.removeBubForRow = function (rowNum) {
    this.rootNode.removeChildByTag(rowNum, true);
};

// delegate
CFSGamePlayScene.prototype.placePaiGowFinished = function () {
    this.placePaiGowRowNum = 999;
    this.enableFunctionMenu();
    this.disablePlacePaiGowMenu();
    this.rootNode.setTouchEnabled(true);
};

CFSGamePlayScene.prototype.turnAroundAllPaiGowRowsFinished = function () {
    this.rootNode.schedule(this.rootNode.onTimerUpdate, 1.0);
    CFSModePaiGowManager.getInstance().highlightCurrentRow();
    this.enableFunctionMenu();
    this.rootNode.setTouchEnabled(true);
};

CFSGamePlayScene.prototype.paiGowDealFinished = function () {
    if (!CFSModePaiGowManager.getInstance().isAnyPairAvialble())  {
        this.rootNode.unschedule(this.rootNode.onTimerUpdate);
        // delay to anounce losing
        var that = this;
        setTimeout(function(){
            CFSModePaiGowManager.getInstance().resetAllPaiGow();
            that.gameFinishedLosing();
        }, 1000);
    }
    else {
        this.enableFunctionMenu();
        this.rootNode.setTouchEnabled(true);
    }
};

CFSGamePlayScene.prototype.showPlacePaiGowArrows = function () {
    this.enablePlacePaiGowMenu();
};

CFSGamePlayScene.prototype.restartGame = function () {
	this.rootNode.setTouchEnabled(false);
    this.timeLabel.setString(MultiLanguageUtil.getLocalizatedStringForKey(STRING_TIMER));
    GameDataManager.resetHintCount();
    this.updateCCBElements();
    this.rootNode.unschedule(this.rootNode.onTimerUpdate);
    this.timeCount = 0;
    this.isGameOver = false;
    this.cfsScoreData.playTimeInSec = 0;
    this.cfsScoreData.hitPercentage = 0.0;
    CFSModePaiGowManager.getInstance().resetAllPaiGow();
    CFSModePaiGowManager.getInstance().initAllPowGowObjects();
    this.disablePlacePaiGowMenu();
    this.runReadySetGoAnimation();
};

CFSGamePlayScene.prototype.resumeGame = function () {
	this.rootNode.setTouchEnabled(true);
    this.enableFunctionMenu();
    this.rootNode.resumeSchedulerAndActions();
    this.isPaused = false;
};

CFSGamePlayScene.prototype.toMainMenu = function () {
	CFSModePaiGowManager.getInstance().resetAllPaiGow();
	this.rootNode.setTouchEnabled(false);
	var scene = cc.BuilderReader.loadAsScene("MainMenuScene.ccbi");
    cc.Director.getInstance().replaceScene(cc.TransitionFade.create(1.0, scene));
};

CFSGamePlayScene.prototype.rowCleared = function (rowNum) {
    this.rootNode.removeChildByTag(ROW_CLEARED_LABEL + rowNum, true);
    var clearLabelSprite = new cc.Sprite();
    clearLabelSprite.initWithSpriteFrameName("levelInfoBg.png");
    clearLabelSprite.setScale(1.2);
    clearLabelSprite.setPosition(cc.p(cc.Director.getInstance().getWinSize().width / 2, PaiGowObject.heightForRow(rowNum)));
    var clearedInfoLabel = cc.LabelTTF.create(MultiLanguageUtil.getLocalizatedStringForKey(STRING_CLEARED), "Verdana-BoldItalic", 12.0);
    clearedInfoLabel.setColor(cc.white);
    clearedInfoLabel.setPosition(cc.p(clearLabelSprite.getContentSize().width / 2, clearLabelSprite.getContentSize().height / 2));
    clearLabelSprite.addChild(clearedInfoLabel);
    this.rootNode.addChild(clearLabelSprite, 999, ROW_CLEARED_LABEL + rowNum);
};

// menu
CFSGamePlayScene.prototype.onPausePressed = function () {
	this.rootNode.setTouchEnabled(false);
	this.disableFunctionMenu();
	this.rootNode.pauseSchedulerAndActions();
	var pauseLayer = cc.BuilderReader.load("PauseLayer.ccbi");
    pauseLayer.controller.delegate = this;
    this.rootNode.addChild(pauseLayer, 9999);
};

CFSGamePlayScene.prototype.onDealPressed = function () {
    if (CFSModePaiGowManager.getInstance().hasPaiGowSelectedInCurrentRow()) {
        if(confirm(MultiLanguageUtil.getLocalizatedStringForKey(STRING_ALERT_DEAL_MESSAGE)) == false) return;
    }
    this.rootNode.setTouchEnabled(false);
    this.disableFunctionMenu();
    CFSModePaiGowManager.getInstance().dealPaiGow();
};

CFSGamePlayScene.prototype.onPlacePaiGowPressed = function (sender) {
    var rowNum = sender.getTag();
    if (rowNum < this.placePaiGowRowNum) {
        if (CFSModePaiGowManager.getInstance().placePaiGow(rowNum)) {
            if (this.placePaiGowRowNum != 999) this.placePaiGowRowNum = rowNum;
            for (var i = rowNum; i <= 5; i ++) this.rootNode.removeChildByTag(80000 + i);
        }
    }
};

CFSGamePlayScene.prototype.onNextPressed = function () {
    if (CFSModePaiGowManager.getInstance().currentRow == 5) return;
    if (CFSModePaiGowManager.getInstance().hasPaiGowSelectedInCurrentRow()) {
        if(confirm(MultiLanguageUtil.getLocalizatedStringForKey(STRING_ALERT_NEXT_MESSAGE)) == false) return;
        CFSModePaiGowManager.getInstance().cancelAllSelectedPaiGow();
    }
    CFSModePaiGowManager.getInstance().moveToNextRow();
    if (!CFSModePaiGowManager.getInstance().isAnyPairAvialble())  {
    	this.rootNode.setTouchEnabled(false);
    	this.disableFunctionMenu();
    	this.rootNode.unschedule(this.rootNode.onTimerUpdate);
    	// delay to anounce losing
    	var that = this;
    	setTimeout(function(){
    		CFSModePaiGowManager.getInstance().resetAllPaiGow();
    		that.gameFinishedLosing();
    	}, 1000);
    }
};

CFSGamePlayScene.prototype.onGetPressed = function () {
    if (GameDataManager.isSoundEffectOn) cc.AudioEngine.getInstance().playEffect("menu.mp3");
    var rs = CFSModePaiGowManager.getInstance().getPaiGowObjectsForRule();
    this.setInfoString(CFSModeRuleManager.getRuleString(rs));
    if (rs == RULETYPE_NONE) {
        this.cfsScoreData.failedHit ++;
        if (GameDataManager.isSoundEffectOn) cc.AudioEngine.getInstance().playEffect("error.mp3");
    }
    else {
        this.cfsScoreData.successfulHit ++;
    }
    if (CFSModePaiGowManager.getInstance().paiGowCountForRow(0) == 32) {
    	this.rootNode.unschedule(this.rootNode.onTimerUpdate);
    	this.rootNode.setTouchEnabled(false);
    	// delay to announce winning
    	var that = this;
    	setTimeout(function(){
    		that.disableFunctionMenu();
    		that.gameFinishedWinning();
    				}, 1000);
    }
};

CFSGamePlayScene.prototype.onHintPressed = function () {
    if (CFSModePaiGowManager.getInstance().showHintPaiGow()) {
        GameDataManager.hintCount --;
        if (GameDataManager.hintCount == 0) this.hintMenuItem.setEnabled(false);
        this.updateCCBElements();
    }
};

// game finish handle
CFSGamePlayScene.prototype.gameFinishedWinning = function () {
	CFSModePaiGowManager.getInstance().resetAllPaiGow();
    var totalHit = this.cfsScoreData.successfulHit + this.cfsScoreData.failedHit;
    this.cfsScoreData.playTimeInSec = this.timeCount;
    //this.cfsScoreData.playDate = new Date();
    this.cfsScoreData.gameLevel = GameDataManager.gameLevel;
    this.cfsScoreData.hitPercentage = (totalHit == 0 ? 0.0 : this.cfsScoreData.successfulHit * 100.0 / totalHit);
    // player name
	var playerName = null;
	var maxLength = 10;
	while (!playerName || (playerName && playerName.length > maxLength)) 
		playerName = window.prompt(MultiLanguageUtil.getLocalizatedStringForKey(STRING_PROMPT_TITLE), 
				   				   MultiLanguageUtil.getLocalizatedStringForKey(STRING_PROMPT_INFO));
	this.cfsScoreData.playerName = playerName;
	// add a waiting label
	var waitingLabel = cc.LabelTTF.create(MultiLanguageUtil.getLocalizatedStringForKey(STRING_SUBMIT_SCORE), "Verdana-BoldItalic", 24);
	waitingLabel.setColor(cc.white);
	waitingLabel.setPosition(cc.p(cc.Director.getInstance().getWinSize().width / 2, cc.Director.getInstance().getWinSize().height / 2));
    this.rootNode.addChild(waitingLabel, 999, WAITING_LABEL);
    // submit score
	var that = this;
	ScoreManager.saveScoreData(that.cfsScoreData, function (cfsScoreData) {
		var gameWinLayer = cc.BuilderReader.load("GameWinLayer.ccbi");
	    gameWinLayer.controller.delegate = that;
		if (!cfsScoreData || !cfsScoreData.id) {
			that.cfsScoreData.id = -1;
			that.rootNode.removeChildByTag(WAITING_LABEL, true);
			gameWinLayer.controller.updateScoreInfo(that.cfsScoreData);
			that.rootNode.addChild(gameWinLayer, 9999);
		}
		else {
			ScoreManager.scoreDataArrayForGameLevel(that.cfsScoreData.gameLevel, function(resultArray) {
																					for (var i = 0; i < Object.size(resultArray); i++) {
																				    	if (resultArray[i].id == cfsScoreData.id) {
																				    		cfsScoreData.rank = i + 1;
																				    		break;
																				    	}
																				    }
																					that.rootNode.removeChildByTag(WAITING_LABEL, true);
																					gameWinLayer.controller.updateScoreInfo(cfsScoreData);
																					that.rootNode.addChild(gameWinLayer, 9999);
																				 }
			);
		}});
};

CFSGamePlayScene.prototype.gameFinishedLosing = function () {
    var gameLoseLayer = cc.BuilderReader.load("GameLoseLayer.ccbi");
    gameLoseLayer.controller.delegate = this;
    this.isGameOver = true;
    this.rootNode.addChild(gameLoseLayer, 9999);
};

File: ClearFiveStages-iOS/ClearFiveStages/Scene/CFSGamePlayScene.m
#import "GameDataManager.h"
#import "SimpleAudioEngine.h"
#import "UIAlertView+Blocks.h"
#import "CFSGamePlayScene.h"
#import "MultiLanguageUtil.h"
#import "CCBReader.h"
#import "CCBAnimationManager.h"
#import "ShowPaiGowLayer.h"
#import "ScoreManager.h"

#define ROW_CLEARED_LABEL   6000

@implementation CFSGamePlayScene
-(void) didLoadFromCCB {
	// add bg
    [self initData];
	[self disableFunctionMenu];
	[self disablePlacePaiGowMenu];
	[[GameDataManager sharedManager] resetHintCount];
	// update elements for multi-language support
	[self updateCCBElements];
    [tabelBgSprite setTexture:[[CCTextureCache sharedTextureCache] addImage:[NSString stringWithFormat:@"tableBg%d.jpg", [GameDataManager sharedManager].tableBgIndex]]];
}

-(void) initData {
	timeCount = 0;
	placePaiGowRowNum = 999;
	shakeStart = [[NSDate alloc] init];
	cfsScoreData = [[CFSScoreData alloc] init];
    cfsScoreData.playTimeInSec = 0;
	cfsScoreData.hitPercentage = 0.0f;
	isPaused = NO;
    isGameOver = NO;
	self.touchEnabled = NO;
	[CFSModePaiGowManager sharedManager].tableLayer = self;	
	[CFSModePaiGowManager sharedManager].delegate = self;
}

#pragma mark - multi-language handle
-(void) updateCCBElements {
    [dealCardButtonLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_DEALCARD]];
    [nextRowButtonLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_NEXTROW]];
    [getCardButtonLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_GETCARD]];
    [hintButtonLabel setString:[NSString stringWithFormat:@"%@(%d)", [MultiLanguageUtil getLocalizatedStringForKey:STRING_SHOWHINT], [GameDataManager sharedManager].hintCount]];
    [timeLabel setString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_TIMER]];
}

#pragma mar - menu
-(void) enableFunctionMenu {
    [dealMenuItem setIsEnabled:YES];
    [getMenuItem setIsEnabled:YES];
    [nextMenuItem setIsEnabled:YES];
    [pauseMenuItem setIsEnabled:YES];
    if ([GameDataManager sharedManager].hintCount > 0) [hintMenuItem setIsEnabled:YES];
}

-(void) disableFunctionMenu {
    [dealMenuItem setIsEnabled:NO];
    [getMenuItem setIsEnabled:NO];
    [nextMenuItem setIsEnabled:NO];
    [hintMenuItem setIsEnabled:NO];
    [pauseMenuItem setIsEnabled:NO];
}

-(void) enablePlacePaiGowMenu {
    for (int rowNum = 1; rowNum <= 5; rowNum ++) {
        if ([[CFSModePaiGowManager sharedManager] isRowCleared:rowNum]) continue;
        CCSprite *arrowSprite = [CCSprite spriteWithSpriteFrameName:@"arrowL.png"];
        arrowSprite.scale = 0.7f;
        CCMenuItemSprite *arrowMenuItem = [CCMenuItemSprite itemWithNormalSprite:arrowSprite selectedSprite:nil target:self selector:@selector(onPlacePaiGowPressed:)];
        arrowMenuItem.tag = rowNum;
        CCMenu *menu = [CCMenu menuWithItems:arrowMenuItem, nil];
        PaiGowObject *paiGowObject = [[CFSModePaiGowManager sharedManager] anyPaiGowForRow:rowNum];
        menu.position = ccp([CCDirector sharedDirector].winSize.width, paiGowObject.pg_position.y);
        [self addChild:menu z:1 tag:80000 + rowNum];
        [menu runAction:[CCRepeatForever actionWithAction:[CCJumpBy actionWithDuration:1.0f position:ccp(0, 0) height:10 jumps:1]]];
    }
}

-(void) disablePlacePaiGowMenu {
    for (int rowNum = 1; rowNum <= 5; rowNum ++) [self removeChildByTag:80000 + rowNum];
}

#pragma mar - animation
-(void)runReadySetGoAnimation {
    CGSize winSize = [CCDirector sharedDirector].winSize;
	CCNode *readySetGoNode = [CCBReader nodeGraphFromFile:@"ReadySetGo.ccbi"];
    readySetGoNode.position = ccp(winSize.width / 2, winSize.height / 2);
    [self addChild:readySetGoNode z:999];
	CCBAnimationManager *animationManager = readySetGoNode.userObject;
	[animationManager setCompletedAnimationCallbackBlock:^(id sender){[self schedule:@selector(timer) interval:1.0f];}];
	[animationManager runAnimationsForSequenceNamed:@"ReadySetGo"];
    [[CFSModePaiGowManager sharedManager] putAllPaiGowsOnTable];
	[[CFSModePaiGowManager sharedManager] turnAroundAllPaiGowRowsWithInterval:0.5f];
}

-(void)setInfoString:(NSString *) info {
    [gameTipLabel stopAllActions];
    [gameTipLabel setString:info];
    [gameTipLabel runAction:[CCSequence actions:
                             [CCFadeIn actionWithDuration:0.1f],
                             [CCBlink actionWithDuration:1.0f blinks:3],
                             [CCFadeOut actionWithDuration:0.5f], nil]];
}

#pragma mark - timer
-(void)timer {
	timeCount++;
	[timeLabel setString:[MultiLanguageUtil getTimeStringFromSecond:timeCount]];
}

#pragma mark - func menu button
-(void)onDealPressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
	if (![[CFSModePaiGowManager sharedManager] hasPaiGowSelectedInCurrentRow]) {
		self.touchEnabled = NO;
		[[CFSModePaiGowManager sharedManager] dealPaiGow];
        [self disableFunctionMenu];
        return;
    }
	NSString* title = [MultiLanguageUtil getLocalizatedStringForKey:STRING_WARNING];
	NSString* message = [MultiLanguageUtil getLocalizatedStringForKey:STRING_ALERT_DEAL_MESSAGE];
	NSString* yesLabel = [MultiLanguageUtil getLocalizatedStringForKey:STRING_CONFIRM_YES];
	NSString* noLabel = [MultiLanguageUtil getLocalizatedStringForKey:STRING_CANCEL_NO];
	RIButtonItem *confirmItem = [RIButtonItem itemWithLabel:yesLabel];
	confirmItem.action = ^ {self.touchEnabled = NO;[self disableFunctionMenu];[[CFSModePaiGowManager sharedManager] dealPaiGow];};
	RIButtonItem *cancelItem = [RIButtonItem itemWithLabel:noLabel];
	cancelItem.action = nil;
	UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title
	                                                    message:message
	                                           cancelButtonItem:cancelItem
	                                           otherButtonItems:confirmItem, nil];
	[alertView show];
	[alertView release];
}

-(void) onPlacePaiGowPressed:(id)sender {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
    int rowNum = ((CCMenuItemSprite *)sender).tag;
    if (rowNum < placePaiGowRowNum) {
	    if ([[CFSModePaiGowManager sharedManager] placePaiGow:rowNum]) {
		    if (placePaiGowRowNum != 999) placePaiGowRowNum = rowNum;
	    	for (int i = rowNum; i <= 5; i ++) {
				[self removeChildByTag:80000 + i];
			}
		}
	}
}
-(void) onGetPressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
	RuleType rs = [[CFSModePaiGowManager sharedManager] getPaiGowObjectsForRule];
	[self setInfoString:[CFSModeRuleManager getRuleString:rs]];
	if (rs == RULETYPE_NONE) {
		cfsScoreData.failedHit ++;
		if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"error.mp3"];
	}
	else {
		cfsScoreData.successfulHit ++;
	}
	if ([[CFSModePaiGowManager sharedManager] paiGowCountForRow:0] == 32) {
        // delay anounce gameFinishedWinning
        [self disableFunctionMenu];
        [self unschedule:@selector(timer)];
        [self performSelector:@selector(gameFinishedWinning) withObject:nil afterDelay:1.0f];
    }
}

-(void)onHintPressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
    if ([[CFSModePaiGowManager sharedManager] showHintPaiGow]) {
	    [GameDataManager sharedManager].hintCount --;
        if ([GameDataManager sharedManager].hintCount <= 0) [hintMenuItem setIsEnabled:NO];
	    [self updateCCBElements];
	}
}

-(void)onPausePressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
    isPaused = YES;
    self.touchEnabled = NO;
    [self pauseSchedulerAndActions];
    [self disableFunctionMenu];
    PauseLayer *pauseLayer = (PauseLayer *)[CCBReader nodeGraphFromFile:@"PauseLayer.ccbi"];
    pauseLayer.delegate = self;
    [self addChild:pauseLayer z:9999];
}

-(void) onNextPressed {
	if ([GameDataManager sharedManager].isSoundEffectOn == YES) [[SimpleAudioEngine sharedEngine] playEffect:@"menu.mp3"];
	if ([CFSModePaiGowManager sharedManager].currentRow == 5) return;
	if (![[CFSModePaiGowManager sharedManager] hasPaiGowSelectedInCurrentRow]) {
		[[CFSModePaiGowManager sharedManager] moveToNextRow];
		if (![[CFSModePaiGowManager sharedManager] isAnyPairAvialble]) {
            [self disableFunctionMenu];
            [self unschedule:@selector(timer)];
            [self performSelector:@selector(gameFinishedLosing) withObject:nil afterDelay:1.0f];
        }
        return;
    }
	NSString* title = [MultiLanguageUtil getLocalizatedStringForKey:STRING_WARNING];
	NSString* message = [MultiLanguageUtil getLocalizatedStringForKey:STRING_ALERT_NEXT_MESSAGE];
	NSString* yesLabel = [MultiLanguageUtil getLocalizatedStringForKey:STRING_CONFIRM_YES];
	NSString* noLabel = [MultiLanguageUtil getLocalizatedStringForKey:STRING_CANCEL_NO];
	RIButtonItem *confirmItem = [RIButtonItem itemWithLabel:yesLabel];
	confirmItem.action = ^ {[[CFSModePaiGowManager sharedManager] moveToNextRow];
        if (![[CFSModePaiGowManager sharedManager] isAnyPairAvialble]) {
            [self disableFunctionMenu];
            [self unschedule:@selector(timer)];
            [self performSelector:@selector(gameFinishedLosing) withObject:nil afterDelay:1.0f];
        }};
	RIButtonItem *cancelItem = [RIButtonItem itemWithLabel:noLabel];
	cancelItem.action = nil;
	UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title
	                                                    message:message
	                                           cancelButtonItem:cancelItem
	                                           otherButtonItems:confirmItem, nil];
	[alertView show];
	[alertView release];
}
#pragma mark - delegate
-(void) turnAroundAllPaiGowRowsFinished {
    [self enableFunctionMenu];
    [[CFSModePaiGowManager sharedManager] highlightCurrentRow];
    self.touchEnabled = YES;
}

-(void) paiGowDealFinished {
	[self enableFunctionMenu];
	self.touchEnabled = YES;
}

-(void) createBubForRow:(int)rowNum hideCount:(int)toHide at:(CGPoint)bubPosition {
	[self removeChildByTag:rowNum cleanup:YES];
	CCSprite *bub = [CCSprite spriteWithSpriteFrameName:@"bub.png"];
	CCLabelTTF *bubInfoLabel = [CCLabelTTF labelWithString:[NSString stringWithFormat:@"×%d", toHide]
											      fontName:@"Verdana-BoldItalic"
											      fontSize:15];
	bubInfoLabel.color = ccBLACK;
	bubInfoLabel.position = ccp(25, 25);
	[bub addChild:bubInfoLabel z:1];
	CCMenuItemSprite *menuItem = [CCMenuItemSprite itemWithNormalSprite:bub selectedSprite:nil target:self selector:@selector(showHiddenPaiGow:)];
    menuItem.scale = 0.7f;
    menuItem.tag = rowNum;  
	CCMenu *menu = [CCMenu menuWithItems:menuItem, nil];
	menu.position = bubPosition;
	[self addChild:menu z:3 tag:rowNum];
	[menu runAction:[CCRepeatForever actionWithAction:[CCJumpBy actionWithDuration:1.5f position:ccp(0, 0) height:6 jumps:1]]];
}

-(void) removeBubForRow:(int)rowNum {
    [self removeChildByTag:rowNum cleanup:YES];
}

-(void) placePaiGowFinished:(int)rowNum {
	placePaiGowRowNum = 999;
    [self enableFunctionMenu];
    [self disablePlacePaiGowMenu];
    self.touchEnabled = YES;
}

-(void) showPlacePaiGowArrows {
	[self enablePlacePaiGowMenu];
}

-(void)showHiddenPaiGow:(id)sender {
	int rowNum = ((CCMenuItemSprite *)sender).tag;
	if ([self getChildByTag:400 + rowNum] == nil) {
		ShowPaiGowLayer *layer = [ShowPaiGowLayer layerWithRow:rowNum];
		[self addChild:layer z:999 tag:400 + rowNum];
	}
}

-(void) gameFinishedWinning {
    [[CFSModePaiGowManager sharedManager] resetAllPaiGow];
	int totalHit = cfsScoreData.successfulHit + cfsScoreData.failedHit;
	cfsScoreData.playTimeInSec = timeCount;
	cfsScoreData.playDate = [NSDate new];
	cfsScoreData.gameLevel = [GameDataManager sharedManager].gameLevel;
	cfsScoreData.hitPercentage = (totalHit == 0 ? 0.0f : (float)cfsScoreData.successfulHit * 100.0f / (float)totalHit);
    cfsScoreData.playerName = [GameDataManager sharedManager].playerName;
    cfsScoreData.rank = [[ScoreManager sharedManager] rankForScore:cfsScoreData];
    if (cfsScoreData.rank <= 10) [[ScoreManager sharedManager] saveHighScoreData:cfsScoreData];
	GameWinLayer *gameWinLayer = (GameWinLayer *)[CCBReader nodeGraphFromFile:@"GameWinLayer.ccbi"];
	[gameWinLayer updateScoreInfo:cfsScoreData];
	gameWinLayer.delegate = self;
	[self addChild:gameWinLayer z:9999];
}

-(void) gameFinishedLosing {
    isGameOver = YES;
    [self disableFunctionMenu];
    [self unschedule:@selector(timer)];
	GameLoseLayer *gameLoseLayer = (GameLoseLayer *)[CCBReader nodeGraphFromFile:@"GameLoseLayer.ccbi"];
	gameLoseLayer.delegate = self;
	[self addChild:gameLoseLayer z:9999];
}

-(void) restartGame {
	self.touchEnabled = NO;
    isGameOver = NO;
	if ([GameDataManager sharedManager].isSoundEffectOn == YES)[[SimpleAudioEngine sharedEngine] playEffect:@"pop.mp3"];
    [timeLabel setString:[NSString stringWithFormat:@"%@0", [MultiLanguageUtil getLocalizatedStringForKey:STRING_TIMER]]];
    [[GameDataManager sharedManager] resetHintCount];
	[self updateCCBElements];
    [self removeAllRowClearedLabel];
    [self setInfoString:@""];
    [self unschedule:@selector(timer)];
    timeCount = 0;
    cfsScoreData.playTimeInSec = 0;
    cfsScoreData.hitPercentage = 0.0f;
    cfsScoreData.successfulHit = 0;
    cfsScoreData.failedHit = 0;
    [[CFSModePaiGowManager sharedManager] resetAllPaiGow];
    [[CFSModePaiGowManager sharedManager] initAllPowGowObjects];
    [self disablePlacePaiGowMenu];
    [self runReadySetGoAnimation];
}

-(void) rowCleared:(int)rowNum {
    [self removeChildByTag:ROW_CLEARED_LABEL + rowNum cleanup:YES];
    CCSprite *clearLabelSprite = [CCSprite spriteWithSpriteFrameName:@"levelInfoBg.png"];
    clearLabelSprite.scale = 1.2f;
    clearLabelSprite.position = ccp([CCDirector sharedDirector].winSize.width / 2, [PaiGowObject heightForRow:rowNum]);
    CCLabelTTF *clearedInfoLabel = [CCLabelTTF labelWithString:[MultiLanguageUtil getLocalizatedStringForKey:STRING_CLEARED]
                                                      fontName:@"Verdana-BoldItalic"
                                                      fontSize:12];
    clearedInfoLabel.color = ccWHITE;
    clearedInfoLabel.position = ccp(clearLabelSprite.contentSize.width / 2, clearLabelSprite.contentSize.height / 2);
    [clearLabelSprite addChild:clearedInfoLabel];
    [self addChild:clearLabelSprite z:999 tag:ROW_CLEARED_LABEL + rowNum];
}

-(void) removeAllRowClearedLabel {
    [self removeChildByTag:ROW_CLEARED_LABEL + 1 cleanup:YES];
    [self removeChildByTag:ROW_CLEARED_LABEL + 2 cleanup:YES];
    [self removeChildByTag:ROW_CLEARED_LABEL + 3 cleanup:YES];
    [self removeChildByTag:ROW_CLEARED_LABEL + 4 cleanup:YES];
    [self removeChildByTag:ROW_CLEARED_LABEL + 5 cleanup:YES];
}

-(void) onEnterTransitionDidFinish {
	if (!isPaused && !isGameOver) {
		[[CFSModePaiGowManager sharedManager] initAllPowGowObjects];
		[self runReadySetGoAnimation];
	}
}

-(void) resumeGame {
	self.touchEnabled = YES;
	[self resumeSchedulerAndActions];
    [self enableFunctionMenu];
	isPaused = NO;
}

-(void) toMainMenu {
    [[CFSModePaiGowManager sharedManager] resetAllPaiGow];
	CCScene* scene = [CCBReader sceneWithNodeGraphFromFile:@"MainMenuScene.ccbi"];
	[[CCDirector sharedDirector] replaceScene:[CCTransitionFadeBL transitionWithDuration:1.0f scene:scene]];
}

-(void)ccTouchesBegan:(NSSet *)touches withEvent:(UIEvent *)event{
	[[CFSModePaiGowManager sharedManager] handleTouch:[touches anyObject]];
}
		 
- (void) dealloc {
	[super dealloc];
}						 
@end
